### ISSUE TITLE
Review Lab n.10 - Tic Tac Toe RL - by Ivan Magistro Contenta s314356

### ISSUE THREAD

**Issue by ivanmag22** [AbdelouahabMoubane/Computational-Intelligence#4]
Hi Abdelouahab!
I want to make a review about your work.

Unfortunately it wasn't easy because there isn't material that explains well the flow of the code (each function has a description, but it was difficult understanding what each of them does) because of the *absence of a clear **README*** that explains well the steps, the configurations and the results, but also for the *absence of plots* (they aren't mandatory, but they could be useful because they are intuitive) and also for the *complexity of the code*, especially for Env_Agent.py (in my opinion the notebook file should explain well each step and the results of the configurations of training and testing).

But I'll try to give my considerations about your work:

- As first thing I want to say that you and your colleague had a good idea of using a **tree** containing the *states of the game*, inspired by the Tic Tac Toe tree that the professor presented to us in order to reduce the number of states that we encounter during the training and the testing of the agent. Especially in your case you used the **Q-Learning**, so I guess that you reduced the number of entries with methods like rotation and simmetry because some situations recur often, but in different positions of the board.
![image](https://github.com/AbdelouahabMoubane/Computational-Intelligence/assets/70893529/bfd41cd8-780d-497b-b443-24096fd724ab)
So the nodes at a each *level* reflect the possible states of a player. So consecutive tree levels refers to different players: one for the first player, another for the second player and again in this manner.
This is not really intuitive, so I recognize your efforts and the results speak for you.

- I've also appreciated the fact that you used a **reward** not only for *wins* and *losses*, but also a negative **reward** for **draws** in order to improve the performances of your agent.

- You've also used good hyper-parameters like learning rate and discount rate. I've also experimented different epsilon for **epsilon-policy** and I found out that higher is the epsilon and better are the results (they grow slowly).

- I understood also that you used two classes, one for the **Agent** and one for the **Environment** that reflect well the concepts of Reinforcement Learning. As I said the **Agent** is a *Q-Learning* one. I see that the agent **trains** with the opponent with *QLearning()* function and **tests** with the opponent with *play_game()* or *play_games()* functions. Moreover I understood that you trained and tested the agent with Q-learning agent or with random player just passing a value to an argument and in this manner you obtained different results. Unfortunately they aren't clear because in the notebook file you should put in different code cells, so I tried to understand just commenting the lines that I wasn't interested in. But also about the print of the game: I didn't understood who was the agent and who the opponent because one game one player starts as first, then in the following game the same player starts as the second. But also agent and opponent weren't associated to specific symbols. So I changed the print in order to obtain the following screens:

  - Q-Agent vs random player
![image](https://github.com/AbdelouahabMoubane/Computational-Intelligence/assets/70893529/e17c4be0-4c74-4257-b74c-3c462883f9de)

  - Q-Agent vs Q-Agent
![image](https://github.com/AbdelouahabMoubane/Computational-Intelligence/assets/70893529/bb50d952-0801-4dc2-9a51-12c5815772c2)
![image](https://github.com/AbdelouahabMoubane/Computational-Intelligence/assets/70893529/c7b67703-5418-48eb-bf8d-d6c3844e5966)

- As regard the **results**, I understood that the agent *improves his performances* if it trains with **Q-learning opponent** instead of training with *random player*. I advice to you to make clear this important found!
Indeed
  - if the agent **trains** with the *random player* and then it is **tested** with *Q-learning opponent* there are 50% of probability that each of them wins
  - if the agent **trains** with the *Q-learning player* and then it is **tested** with *Q-learning opponent* they always have a draw and the agent never lose with *random player*, while if trained with random player it could lose with it.

- Unfortunately I didn't understood if with Q-agent vs Q-opponent there is one **Q-table** for each of them of one only for the agent.

For the rest I tell you that you've done a great work and I really appreciated it, although these considerations. I hope that my advices could be useful for you!
Have a good work and good luck for the exam! :)
